src/redprl/refiner_types.fun

```diff
@@ -15,6 +15,76 @@ struct
   infix 2 >> >: >:? >:+ $$ $# // \ @>
   infix orelse_
 
+  (* Rules in a type should be ordered as follows:
+   *
+   * EqType: the type formation rule.
+   * EqX: the introduction rule for the constructor X of a positive type.
+   * True: the introduction rule of a negative type.
+   * Eta: the eta rule, if any.
+   * Elim/FullElim: the elimination rule (see the long comment below).
+   * EqElim/EqX: structural equality for eliminators.
+   *   We use EqX if the eliminator has a well-known name X.
+   *   For example, we have EqApp for DFun and Path, and EqProj for Record.
+   * (others): other special rules for this type.
+   *)
+
+  (* Note that there are two categories of elimination rules for positive types.
+   * The main issue is, given the sequent `\Gamma, x:A, \Delta >> J`, how should
+   * we eliminate `x:A` for a positive type `A`?
+   *
+   * The first category, `FullElim`, works on almost arbitrary J by considering
+   * all elements in A, and is justified by the semantics of open judgments.
+   * For example, `tt` and `ff` are the only elements in Bool, and so to show
+   *
+   * `\Gamma, x:Bool, \Delta >> J`
+   *
+   * it is sufficient to consider two subgoals:
+   *
+   * 1. `\Gamma, x:bool, \Delta[tt/x] >> J[tt/x]`
+   * 2. `\Gamma, x:bool, \Delta[ff/x] >> J[ff/x]`
+   *
+   * (The `x:Bool` is kept to follow the Nuprl style. It might not be needed.)
+   * In RedPRL, we restrict J to judgments where there is a sensible realizer
+   * that can be easily reconstructed from the subgoals.
+   *
+   * Unfortunately, it seems difficult to have this powerful elimination principle
+   * for every type we have in RedPRL, because a type could have an infinite number
+   * of "cases" or demand relations between these cases. Take the circle `S1`
+   * for example: it has elements `base`, `loop` and `fcom`; there are infinitely
+   * many different shapes of `fcom` and thus infinitely many cases (unless our LF
+   * can deal with them uniformly), and there are (infinitely) many equalities
+   * between these elements due to cubicality.
+   *
+   * Another example is the natural number type `nat` which has elements `zero`,
+   * `(succ zero)`, ...; this also has infinitely many cases. It is imaginable
+   * that we can generate two "higher-order subgoals" in some higher-order LF:
+   *
+   * 1. `\Gamma, x:nat, \Delta[zero/x] >> J[zero/x]`
+   * 2. if `\Gamma, x:nat, \Delta[#n/x] >> J[#n/x]` is provable for some
+   *    numeral `n`, then `\Gamma, x:nat, \Delta[(succ #n)/x] >> J[(succ #n)/x]`
+   *    is provable.
+   *
```

Comment:
I think this might be the only thing that could make sense if it ever made sense, but I support the idea of removing them from the comment unless we know concretely what kind of "higher-order LF" we want/have.